ace.d.ts

```diff
@@ -0,0 +1,766 @@
+
+  export interface Range {
+    start: Point;
+    end: Point;
+
+    isEqual(range: Range): boolean;
+    toString(): string;
+    contains(row: number, column: number): boolean;
+    compareRange(range: Range): number;
+    comparePoint(p: Point): number;
+    containsRange(range: Range): boolean;
+    intersects(range: Range): boolean;
+    isEnd(row: number, column: number): boolean;
+    isStart(row: number, column: number): boolean;
+    setStart(row: number, column: number): void;
+    setEnd(row: number, column: number): void;
+    inside(row: number, column: number): boolean;
+    insideStart(row: number, column: number): boolean;
+    insideEnd(row: number, column: number): boolean;
+    compare(row: number, column: number): number;
+    compareStart(row: number, column: number): number;
+    compareEnd(row: number, column: number): number;
+    compareInside(row: number, column: number): number;
+    clipRows(firstRow: number, lastRow: number): Range;
+    extend(row: number, column: number): Range;
+    isEmpty(): boolean;
+    isMultiLine(): boolean;
+    clone(): Range;
+    collapseRows(): Range;
+    toScreenRange(session: EditSession): Range;
+    moveBy(row: number, column: number): void;
+  }
+
+  export interface EditorOptions {
+    selectionStyle: string;
+    highlightActiveLine: boolean;
+    highlightSelectedWord: boolean;
+    readOnly: boolean;
+    copyWithEmptySelection: boolean;
+    cursorStyle: 'ace' | 'slim' | 'smooth' | 'wide';
+    mergeUndoDeltas: true | false | 'always';
+    behavioursEnabled: boolean;
+    wrapBehavioursEnabled: boolean;
+    autoScrollEditorIntoView: boolean;
+    keyboardHandler: string;
+    value: string;
+    session: EditSession;
+
+    // Pass-through to VirtualRenderer
+    hScrollBarAlwaysVisible: boolean;
+    vScrollBarAlwaysVisible: boolean;
+    highlightGutterLine: boolean;
+    animatedScroll: boolean;
+    showInvisibles: boolean;
+    showPrintMargin: boolean;
+    printMarginColumn: number;
+    printMargin: boolean | number;
+    fadeFoldWidgets: boolean;
+    showFoldWidgets: boolean;
+    showLineNumbers: boolean;
+    showGutter: boolean;
+    displayIndentGuides: boolean;
+    fontSize: number;
+    fontFamily: string;
+    maxLines: number;
+    minLines: number;
+    scrollPastEnd: boolean;
+    fixedWidthGutter: boolean;
+    theme: string;
+    hasCssTransforms: boolean;
+
+    // Pass-through to MouseHandler
+    scrollSpeed: number;
+    dragDelay: number;
+    dragEnabled: boolean;
+    focusTimeout: number;
+    tooltipFollowsMouse: boolean;
+
+    // Pass-through to EditSession
+    firstLineNumber: number;
+    overwrite: boolean;
+    newLineMode: NewLineMode;
+    useWorker: boolean;
+    useSoftTabs: boolean;
+    navigateWithinSoftTabs: boolean;
+    tabSize: number;
+    wrap: boolean | number;
+    indentedSoftWrap: boolean;
+    foldStyle: 'markbegin' | 'markbeginend' | 'manual';
+    mode: string;
+  }
+
+  export interface EditSessionOptions {
```

Comment:
Is there a way to not duplicate this? so that we have  EditorOptions, EditSessionOptions, RendererOptions, MouseHandlerOptions, and editor accepts any of these?

Reply:
the editor only accepts an `EditorOptions` from what i can see. but it inherits _some_ (by pass-through) of the others. i could just make it extend the other few interfaces and ignore the fact that not all of them get forwarded through?

Reply:
Editor is meant to inherit all options of the others.

Reply:
ah cool ok, so if i see any that don't, i'll add them into the define call